src/fastcs/transport/graphQL/graphQL.py

```diff
@@ -44,9 +46,58 @@ def run(self, options: GraphQLServerOptions | None = None) -> None:
         )
 
 
+class GraphQLAPI:
+    """A Strawberry API built dynamically from a Controller"""
+
+    def __init__(self, controller: BaseController):
+        self.queries: list[StrawberryField] = []
+        self.mutations: list[StrawberryField] = []
+
+        api = _get_single_mapping(controller)
+
+        for attr_name, attribute in api.attributes.items():
+            match attribute:
+                # mutation for server changes https://graphql.org/learn/queries/
+                case AttrRW():
+                    self.queries.append(
+                        strawberry.field(_wrap_attr_get(attr_name, attribute))
+                    )
+                    self.mutations.append(
+                        strawberry.mutation(_wrap_attr_set(attr_name, attribute))
+                    )
+                case AttrR():
+                    self.queries.append(
+                        strawberry.field(_wrap_attr_get(attr_name, attribute))
+                    )
+                case AttrW():
+                    self.mutations.append(
+                        strawberry.mutation(_wrap_attr_set(attr_name, attribute))
+                    )
+
+        for cmd_name, method in api.command_methods.items():
+            self.mutations.append(
+                strawberry.mutation(_wrap_command(cmd_name, method.fn, controller))
+            )
+
+        for sub_controller in controller.get_sub_controllers().values():
```

Comment:
Yeah it could just be a function now, but I do like having the queries and mutations in a container rather than passing tuples around and having to make sure which way around the items are. And if we have that just as a dataclass, then this method is effectively its constructor and you would never create one manually.

If the function was split up then `GraphQLServer` would have to call those lower level functions and combine the results, which exposes a some implementation details. I was actually considering encapsulating more into `GraphQLAPI` and giving it a method to create the schema, so the top level would just be

```python
        api = GraphQLAPI(self._controller)
        app = GraphQL(api.create_schema())
```

I don't think the function is too long, and I think it is pretty easy to understand what it is doing at a glance, but I could be convinced to have `_process_attributes`, `_process_commands`, `_process_children` as methods on the class that are called in `__init__`. Do you think that would be better?

Reply:
Yes, I like your proposals - this is looking very clean. Good work